### PR TITLE
agent: stop endpoints in parallel on exit

### PR DESCRIPTION
**Pull request by jaffcheng**
agent: stop endpoints in parallel on exit

Endpoints are stopped on exit signals in an agent cleanup function.
This patch does this in goroutines to speed it up, reduces the probability
of agent exiting timeout, that is, reduces the possibility of pod network
disconnection caused by interrupted regeneration.

Related: #15446
Signed-off-by: Jaff Cheng jaff.cheng.sh@gmail.com

```release-note
Wait for endpoints to be stopped on agent shutdown
```

### CODE REVIEW

**Comment by pchaigno** [comment]
test-me-please

File: pkg/endpoint/endpoint.go [hunk 1]
```diff
@@ -2308,3 +2308,11 @@ func (e *Endpoint) setDefaultPolicyConfig() {
 func (e *Endpoint) GetCreatedAt() time.Time {
 	return e.createdAt
 }
+
+// StopEventQueue stops the event queue of this endpoint.
+func (e *Endpoint) StopEventQueue() {
+	e.eventQueue.Stop()
+	e.getLogger().Debug("eventQueue stopped")
+	e.eventQueue.WaitToBeDrained()
+	e.getLogger().Debug("eventQueue drained")
```
Comment (pchaigno) [inline pkg/endpoint/endpoint.go]: We already do this in `Endpoint.Stop()` and we already call that function on agent shutdown: https://github.com/cilium/cilium/blob/141cd637d8fc028e46d6912bb6d852ad89779377/daemon/cmd/daemon.go#L407-L412

Why is that not enough? Maybe it's just a matter of adding a WaitGroup for those stops?

Comment (jaffcheng) [reply pkg/endpoint/endpoint.go]: Oops, sorry I didn't realize we already do this in master.
Yes, I'll add a WaitGroup to the existing function.

**Comment by jaffcheng** [comment]
I guess this should be a minor change now.

**Comment by aditighag** [comment]
This is a nice improvement, but I don't see how this fixes the issue you've tagged. 

>This happens when cilium-agent is killed (crash/upgrade/operational restart) during endpoint regeneration, more specifically, during synchronizeDirectories function call.

If an agent crashes or killed, then wait groups are not going to add any protection anyway. 

>In the scenario of RegenerateWithoutDatapath (e.g. one or more identities created or deleted), next dir is empty at this step. So after the following code, we have an empty state dir 1111 and a complete backup 1111_stale
If the agent stops at this point, after the agent restarts, we lose track of this endpoint, tail call maps will get GCed, and the pod will lose connectivity.

The way we take backup across 2 steps (`os.Rename` is an atomic call) should let us restore from the backup upon agent restart.

**Comment by jaffcheng** [comment]
> This is a nice improvement, but I don't see how this fixes the issue you've tagged.

Yes, I've commented in the issue: https://github.com/cilium/cilium/issues/15446#issuecomment-807952086, also changed the Fixes tag to Related in commit msg.



> The way we take backup across 2 steps (`os.Rename` is an atomic call) should let us restore from the backup upon agent restart. 🤔 I don't quite follow why "tail call maps will get GC'd" as they would still being referenced in BPF programs.


I think we just ignored backups while restoring endpoints. I'm not sure if this is intended or just an overlook.
https://github.com/cilium/cilium/blob/52ec2f6db8fd8d52414ec1360c56e33f39109e46/pkg/endpoint/restore.go#L170-L178

**Comment by pchaigno** [comment]
test-me-please

File: daemon/cmd/daemon.go [hunk 2]
```diff
@@ -406,9 +407,20 @@ func NewDaemon(ctx context.Context, cancel context.CancelFunc, epMgr *endpointma
 	}
 	// Stop all endpoints (its goroutines) on exit.
 	cleaner.cleanupFuncs.Add(func() {
-		for _, ep := range d.endpointManager.GetEndpoints() {
-			ep.Stop()
+		var wg sync.WaitGroup
```
Comment (aanm) [inline daemon/cmd/daemon.go:411]: Might be worth adding a Info message here informing the user that Cilium is waiting for the endpoints' go routines to be stopped.

**Review by jrajahalme — APPROVED** [review]

**Review by pchaigno — APPROVED** [review]

**Comment by pchaigno** [comment]
test-me-please

**Comment by jaffcheng** [comment]
> > I think we just ignored backups while restoring endpoints. I'm not sure if this is intended or just an overlook.
> 
> We should probably fix that.

I'll try to fix that in a separate PR.

**Review by aanm — APPROVED** [review]

**Comment by jaffcheng** [comment]
Rebased

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-1.16-netnext

**Pull request merged**